### PR TITLE
Fuse Multiple DEFINE Types Into Single Evaluation Loop

### PR DESCRIPTION
**Pull request by bska**
This commit switches the implementation of eval_define() into a single loop over `UDQConfig::input_index` instead of multiple loops over subsets of the same input index.  While this is more complex logic, it saves forming multiple intermediate quantities that will for the most part be discarded.

This is also in preparation of adding support for segment level UDQs, which would necessitate another loop in the previous control structure.

While here, also switch to using structured bindings where appropriate and pruning a few other intermediate objects.

### CODE REVIEW

**Comment by bska** [comment]
jenkins build this please

File: src/opm/input/eclipse/Schedule/UDQ/UDQConfig.cpp [hunk 1]
```diff
@@ -480,25 +469,30 @@ namespace Opm {
...
+                continue;
             }
-        }
 
-        for (const auto& def : this->definitions(UDQVarType::FIELD_VAR)) {
-            if (udq_state.define(def.keyword(), def.status())) {
-                auto field_udq = def.eval(context);
-                context.update_define(report_step, def.keyword(), field_udq);
+            auto def_pos = this->m_definitions.find(keyword);
+            if ((def_pos == this->m_definitions.end()) || // No such def
```
Comment (blattms) [inline src/opm/input/eclipse/Schedule/UDQ/UDQConfig.cpp:488]: Is this semantic change intentional?
Previously an exception would have been thrown if there  is no entry in definitions. Now we silently continue

**Review by blattms — COMMENTED** [review]
This looks good (even though I always struggle with bit-operations, but that is probably just me). Just one tiny question about a hidden change in semantics,

File: src/opm/input/eclipse/Schedule/UDQ/UDQConfig.cpp (hunk 1 above)
Comment (bska) [reply src/opm/input/eclipse/Schedule/UDQ/UDQConfig.cpp]: > Is this semantic change intentional?

Thanks for noticing.  That wasn't an intended part of the change.  I've reinstated the exception, but this time with a (hopefully) more informative diagnostic message than before.  Previously we would have gotten something like
```
Internal error: _Map_base::at
```
and we already have enough of those, e.g., OPM/opm-simulators#3806.

**Comment by bska** [comment]
jenkins build this please

**Comment by blattms** [comment]
Perfect. thanks.

**Pull request merged**